### PR TITLE
Update build-all test for null-safe template

### PR DESCRIPTION
**Pull request by stuartmorgan**
Flutter master now creates NNBD code when running 'flutter create', so
the generated pubspec for build_all needs to use a compatible SDK
version. This updates from 2.0.0 to 2.12.0.

Also includes a test for this, which involved setting up tests for the
file, and doing some refactoring to make the command testable. As a
result, this fixes https://github.com/flutter/flutter/issues/61049
(although more test backfill is needed).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I signed the [CLA].
- [ ] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

File: script/build_all_plugins_app.sh [hunk 1]
```diff
@@ -53,7 +59,7 @@ fi
 
 for version in "${BUILD_MODES[@]}"; do
   echo "Building $version..."
-  (cd $REPO_DIR/all_plugins && flutter build $@ --$version)
+  (cd $REPO_DIR/all_plugins && flutter build $@ --$version --no-sound-null-safety)
```
[the remark this answers is not in the document]
Comment (ditman) [reply script/build_all_plugins_app.sh:62]: Yep, I'll update mine once this lands. (For cross-linking purposes, this is what's supposed to be updated: https://github.com/flutter/plugins/pull/3726)

**Review by ditman — APPROVED** [review]
LGTM, thanks for the test!

**Comment by stuartmorgan** [comment]
Landing on red since all PRs are broken on this.

File: script/build_all_plugins_app.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (ditman) [reply script/build_all_plugins_app.sh:62]: (Done: https://github.com/flutter/plugins/pull/3726/commits/479a543febbf120c3f10630dfe4edaa2e1c7d36c )

**Pull request merged**